src/event.rs

```diff
@@ -1,64 +1,60 @@
+#![allow(non_upper_case_globals)]
 use enet_sys::{
     ENetEvent, _ENetEventType_ENET_EVENT_TYPE_CONNECT, _ENetEventType_ENET_EVENT_TYPE_DISCONNECT,
     _ENetEventType_ENET_EVENT_TYPE_NONE, _ENetEventType_ENET_EVENT_TYPE_RECEIVE,
 };
 
 use crate::{Packet, Peer};
 
-/// This enum represents an event that can occur when servicing an `EnetHost`.
-///
-/// Also see the official ENet documentation for more information.
+/// This struct represents an event that can occur when servicing an `Host`.
 #[derive(Debug)]
-pub enum Event<'a, T> {
-    /// This variant represents the connection of a peer, contained in the only field.
-    Connect(Peer<'a, T>),
-    /// This variant represents the disconnection of a peer, either because it was requested or due to a timeout.
-    ///
-    /// The disconnected peer is contained in the first field, while the second field contains the user-specified
-    /// data for this disconnection.
-    Disconnect(Peer<'a, T>, u32),
-    /// This variants repersents a packet that was received.
+pub struct Event<'a, T> {
+    /// The peer that this event happened on.
+    pub peer: &'a mut Peer<T>,
+    /// The type of this event.
+    pub kind: EventKind,
+}
+
+/// The type of an event.
+#[derive(Debug)]
+pub enum EventKind {
+    /// Peer has connected.
+    Connect,
+    /// Peer has disconnected.
+    //
+    /// The data of the peer will be dropped on the next call to Host::service or when the structure is dropped.
+    Disconnect {
+        /// The data associated with this event. Usually a reason for disconnection.
+        data: u32,
+    },
+    /// Peer has received a packet.
     Receive {
-        /// The `Peer` that sent the packet.
-        sender: Peer<'a, T>,
-        /// The channel on which the packet was received.
+        /// ID of the channel that the packet was received on.
         channel_id: u8,
-        /// The `Packet` that was received.
+        /// The received packet.
         packet: Packet,
     },
 }
 
 impl<'a, T> Event<'a, T> {
-    pub(crate) fn from_sys_event<'b>(event_sys: &'b ENetEvent) -> Option<Event<'a, T>> {
-        #[allow(non_upper_case_globals)]
-        match event_sys.type_ {
-            _ENetEventType_ENET_EVENT_TYPE_NONE => None,
-            _ENetEventType_ENET_EVENT_TYPE_CONNECT => {
-                Some(Event::Connect(Peer::new(event_sys.peer)))
-            }
-            _ENetEventType_ENET_EVENT_TYPE_DISCONNECT => Some(Event::Disconnect(
-                Peer::new(event_sys.peer),
-                event_sys.data,
-            )),
-            _ENetEventType_ENET_EVENT_TYPE_RECEIVE => Some(Event::Receive {
-                sender: Peer::new(event_sys.peer),
+    pub(crate) fn from_sys_event(event_sys: &ENetEvent) -> Option<Event<'a, T>> {
+        if event_sys.type_ == _ENetEventType_ENET_EVENT_TYPE_NONE {
+            return None;
+        }
+
+        let peer = Peer::new_mut(unsafe { &mut *event_sys.peer });
```

Comment:
This is not safe I think. I can just call this method twice with the same `event_sys` and I will have two mutable references to `Peer`. Why do you want to store a `&mut Peer` instead of just a `&Peer` in the `Event` struct?